CIPs/CIP-85/cip-85.md

```diff
@@ -0,0 +1,193 @@
+- `slicesCount`: the total number of slices the collection contains
+- `first`: DocID of the first slice
+- `last`: DocID of the last slice
+
+```js
+{
+  $schema: 'http://json-schema.org/draft-07/schema#',
+  title: 'MyCollection',
+  type: 'object',
+  $id: 'ceramic://appendCollection',
+  properties: {
+    sliceSize: { type: 'integer', minimum: 10, maximum: 256 },
+    slicesCount: { type: 'integer', minimum: 1 },
+    first: { type: 'string', $id: 'ceramic://doc', $ceramicSchema: '<slice schema ID>', maxLength: 150 },
+    last: { type: 'string', $id: 'ceramic://doc', $ceramicSchema: '<slice schema ID>', maxLength: 150 },
+  },
+  required: ['sliceSize', 'slicesCount', 'first', 'last']
+}
+```
+
+#### CollectionSlice schema
+
+The `CollectionSlice` schema must be an `object` with a [`$id` field](https://json-schema.org/draft/2019-09/json-schema-core.html#rfc.section.8.2.2) value of `ceramic://collectionSlice` and the following properties:
+
+- `contents`: array with a `maxItems` value matching the `AppendColelction` `sliceSize` property and defining the items schemas, that must include `{ type: 'null' }` in order to support removals
+- `next`: DocID of the next slice (if the slice is not the last one)
+- `prev`: DocID of the previous slice (if the slice is not the first one)
+
+```js
+{
+  $schema: 'http://json-schema.org/draft-07/schema#',
+  title: 'MyCollectionSLice',
+  type: 'object',
+  $id: 'ceramic://collectionSlice',
+  properties: {
+    contents: {
+      type: 'array',
+      maxItems: 256,
+      minItems: 1,
+      items: {
+        oneOf: [{ type: 'object', ... }, { type: 'null' }],
+      },
+    },
+    next: { type: 'string', $id: 'ceramic://doc', $ceramicSchema: '<slice schema ID>', maxLength: 150 },
+    prev: { type: 'string', $id: 'ceramic://doc', $ceramicSchema: '<slice schema ID>', maxLength: 150 },
+  },
+  required: ['contents'],
+}
+```
+
+### Algorithms
+
+#### First insertion
+
+> This flow assumes a prerequisite check that the `AppendCollection` document has not been created yet.
+
+1. Create the `CollectionSlice` document with the `contents` array containing the item to insert.
+1. Create the `AppendCollection` document with a `slicesCount` of `1`, and the `first` and `last` values containing the DocID of the created `CollectionSlice`.
+
+#### Other insertions
+
+1. Load the `AppendCollection` document.
+1. Load the `CollectionSlice` document from the `last` field of the `AppendCollection`.
+1. Check the length of the `contents` array of the `CollectionSlice`:
+
+- If it is lower than the `sliceSize` value of the `AppendCollection`, add the item to the `contents` array.
+- If it is equal to the `sliceSize` value:
+  1. Create a new `CollectionSlice` document with the `contents` array containing the item to insert and the `prev` value containing the DocID of the previous slice.
+  1. Update the previous slice to set the `next` value to the DocID of the newly created slice.
```

Comment:
I'm a bit concerned about the lack of atomicity between this step and the next one.  I worry that conflicting updates, or a crash in the middle of this process, could leave the `next` pointer from the `CollectionSlice` out of sync with the `last` pointer from the `AppendCollection`.

I'm not totally sure what to do about it though.  The `next` pointer is important to have within the `CollectionSlice` so that scanning forward works efficiently.

I wonder, could we take advantage of deterministic documents to make the `CollectionSlice` docIds knowable deterministically from the `AppendCollection` docId and the slice index?

@oed @PaulLeCam

Reply:
Deterministic slice IDs would also have the nice advantage of allowing binary search and other types of scans over the collection other than strict front-to-back or back-to-front

Reply:
Yes good point about the atomicity, my expectation was that these 2 calls would be done in parallel and if either fails there's a retry process, but obviously that's up to the implementation logic and that doesn't really solve the problem of being atomic.

Deterministic slice IDs is a great idea! That would prevent the need for a `next` reference so adding a new slice would only need to update the `AppendCollection`, and that could speedup loading slices as well as this could be done in parallel rather than sequentially.
How do you see it working in the schema please? Just thinking about it a bit I think the deterministic slice should be something like `{ collection: '<collection docID>', sliceNumber: number, contents: [] }` and then adding the content and `prev` reference when updating, do you think that works?

Reply:
Yeah, that looks great for how the genesis commit is generated.

I actually think with this approach you don't need the `prev` reference in `CollectionSlice` since you can always get the previous slice docID by subtracting 1 from the current sliceNumber.  I also think we can then remove the `first` and `last` reference from `AppendCollection`, and just use the `slicesCount` to find the current "last" slice.

Reply:
Yes good point! I updated the CIP in https://github.com/ceramicnetwork/CIP/pull/91 if you can have a look please?